### DISCUSSION TITLE
Character Import Improvements

### DISCUSSION THREAD

**Discussion by amzn-tommy** [o3de/o3de#2056] [General]
Hello! I'm an engineer on the Atom team, working on skinning and the mesh pipeline for Atom. So I've spent some time importing characters :)

I'm hoping we can make a communal effort to improve the character import process over the coming months, and get more things working out of the box with default settings for importing characters. While the current pipeline works well for people who have built up more time and experience developing a process for what works and what doesn't (which is fair for an engine targeting AAA game development), I don't want character importing to become something that blocks adoption for first time users/evaluators who want to drop a character into the engine, and have it 'just work'. A few areas I'd like to improve are: Support for content from other sources, fewer restrictions on what can be imported, better support for scene rules, and better automated testing for imported characters.

Between providing feedback, reporting issues, guiding the prioritization, and contributing back, a community effort will help bring this all together.

### Content from other sources
In theory, a character is just a character, no matter where it came from. But in practice, different dcc tools export fbx files in different ways, so we've found that not all of them work out of the box.

Maya:
Some context - most of Lumberyard's bigger customers have used a Maya based workflow for creating characters in the past, so exporting an fbx from Maya has been one of the better supported golden paths for importing characters. With the switch from using the FBX SDK to Assimp for importing an fbx, we ran into several regressions, in particular with complex rigs like the one used for the ['Old World' teaser](https://www.youtube.com/watch?v=ebr8t_mz8p8), which the team has been working through. 

Blender:
Anecdotally, fbx files exported from Blender seem to play nicer with Assimp than those exported from Maya. But because we've been working closely with Apocalypse prior to the public developer preview, a lot of what we've been using to stress test the new pipeline has been from Maya, so we have less coverage here.

Mixamo:
Mixamo is great, and there have been some ad hoc efforts to get Mixamo assets working with the engine. But the content as it comes directly out of Mixamo does not work unmodified. [Lumbermixalot ](https://github.com/lumbermixalot/lumbermixalot) is a Blender plugin that was created to make the coordinate system changes necessary to import content from Mixamo, but ideally we'd like to see these assets work successfully via the Asset Processor without needing to insert an intermediate step. Adobe is a parter of Open3D, so it would be wonderful if we could get direct support from them, but I don't have any insight into their partnership, so I don't know if Mixamo is something they intend to support or if their focus will be on their other products.

Daz3D:
When I saw the AWS blogpost from Tafi about being an O3DE partner, I took a couple of hours to try importing a character from Daz3D. While I was able to get a character mesh imported (as long as it didn't have any props), the animations did not work by default (looked like another coordinate system issue).

### Fewer restrictions on what can be imported
We've seen some limitations in the workflows used by Apocalypse for the Old World teaser, which they've been able to work around in content, but would have a better experience if they didn't have to work around them
- Mirroring a group or controller doesn't work
-- the controller will have a negative scale/negative transform node, which the current o3de pipeline doesn't handle (or at least didn't as of a few months ago)
-- beneficial for mirroring in Maya (symmetric parts of the skeleton)
- Having blend shapes on multiple sub-meshes doesn't work (https://github.com/o3de/o3de/issues/3037)
- Having sub-meshes that lack skinning weights causes the character to fail to import
-- this in particular happens when first trying to import work in progress assets into the engine
- Having too many vertex streams causes a model to fail to import
-- https://github.com/o3de/o3de/issues/1987

### Better support for scene rules
By default, if an fbx has an animation in it in addition to the mesh/skeleton data, it will not produce a .actor that can be used with the actor component. You have to add it manually in the rules, or export the animation separately and keep the animation out of the fbx that produces the actor. This was done to support the most common usage among our customers which were exporting mesh and skeletal data in a separate file from the animations. This is a good workflow to follow, but for new users it requires some tribal knowledge.

Additionally, we've found that several import issues can be caused or fixed by tweaking the Assimp settings, which are currently hard coded. For example, Assimp mesh optimization was breaking morph targets, so it was temporarily disabled. We'd like to make these settings data driven, but I'd also like to see if we can pick reasonable default rules based on the content of the scene. For example, if it would be better if we could disable Assimp mesh optimization only if we find that it has morph targets. This can be tricky because we want to avoid double processing the asset, but it might be worthwhile to have a two-pass processing for a scene that does not yet have a corresponding .scenesettings file.

On a similar note, we have a coordinate system rule which can be used to resolve some of the issues seen when importing characters, but ideally we could generate a smart set of defaults that make assets coming from software with varying coordinate systems 'just work' out of the box without manual intervention.

### Better Automated Testing
We have some complex assets that can't be included in the AutomatedTesting project because they are either too large, or they are work in progress production assets from our customers that can't be shared publicly. Ideally, we'd like to boil each of these down into a simpler asset that exercises the edge cases in the import pipeline. If you are a content creator, this is one place we could really use your help! If you encounter a character that does not import properly, and you are able to break it down to the simplest possible asset that exhibits the same behavior, that would help us tremendously. It would make it easier to debug and fix the root cause, and also provide us with something we could use in test automation moving forward to ensure it does not regress.

These are just some of my thoughts. I'd like to hear more from our customers on where we can improve. We can use this thread to discuss the topic more with the o3de community.

**Comment by amzn-rhhong** [comment]
yo @amzn-tommy this is great discussion to have, just here to chip in my two cents:

**Area I think could use improvement**
**Retargeting**: https://docs.aws.amazon.com/lumberyard/latest/userguide/animation-editor-retargeting-animations.html od3e inherited lumberyard's rather tedious retargeting workflow that requires user to add bindpose to first frame and then uses the motion range rule to chop out motions.
**LOD**: Many items have been moved to post GA. The whole skeleton / mesh LOD setup is still a bit unintuitive.
**Better warning/feedback in general** - in AP, ly-editor and animation editor. For example root bone transform not zeroed-out should be a warning.
**Atom Renderer integration to animation editor**: Will remove tech debt, what-you-see-is-what-you-get in animation editor, leverage some of the atom tech for improvement

re- The Maximo importing issue: I'd like to know what's causing the problem. A couple weeks ago I've been playing around with Maximo with rin/burt with maximo animation. I used the auto rigger to generate skeleton for the character and I have not running into any problem. 

Echo on the automated testing part. Also I'd like to point out that we already have a few unit test in-place to test a variety parts of pipeline, from mesh optimizer to actor builder. I think a higher-level, last resort type of sanity check level that contains a bunch of character is a very nice to have.

**Comment by WashedUpStudio** [comment]
I have spent a long while trying to import some characters. I have so far two assets that are characters with skeletons, and one which is a robot imports correctly every time. I also have an astronaut which does not import correctly anymore. As soon as a public release of o3de came out I got a build made and imported my astronaut first. No issues that were not my fault and it had a skeleton and was working perfect. A few weeks passed and now I can no longer get my character in with a skeleton. This one issue will hopefully be figured out since amzn-tommy can help out with it in time. So from my perspective, the fbx importer is working well and is nearly an import and use process, with the exception of this one issue. If we can figure out what this issue is and get it resolved, I can tell others that a blender to o3de workflow for characters works as expected with no issues on the engine side.

**Comment by amzn-jillich** [comment]
Related issue: Skeletons are imported incorrectly due to inaccessible bind pose transforms for joints that do not skin at least one vertex (https://github.com/o3de/o3de/issues/2687)

**Reply by amzn-jillich** [reply]
Created an issue for "Having blend shapes on multiple sub-meshes doesn't work": https://github.com/o3de/o3de/issues/3037

**Comment by sysuNie** [comment]
More restrictions on what can be imported:
1. FBX model parsing with multiple animations is not supported
2. VertexStream 'xxx' does not match the expected vertex count. This typically means multiple sub-meshes have mis-matched vertex stream layouts (such as one having more uv sets than the other) but are assigned the same material in the dcc tool so they were merged.

**Reply by amzn-tommy** [reply]
1 sounds like a regression, I know that has worked before. I will try to reproduce it on my end.

2 is a known issue which impacts static meshes as well. I am aware of how to fix that, but it has not been priority thus far since it can be fixed in conent. https://github.com/o3de/o3de/issues/7588
